File: 1110.delete-nodes-and-return-forest.java

// @leet start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode() {} TreeNode(int val) { this.val = val; } TreeNode(int val, TreeNode left,
 * TreeNode right) { this.val = val; this.left = left; this.right = right; } }
 */
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Set;
import java.util.stream.Collectors;

class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        Set<TreeNode> forest = new HashSet<>();
        Set<Integer> unwanted = Arrays.stream(to_delete).boxed().collect(Collectors.toSet());

        Queue<TreeNode> queue = new LinkedList<>();

        if (!unwanted.contains(root.val)) {
            forest.add(root);
        }
        queue.add(root);

        while (!queue.isEmpty()) {
            TreeNode curr = queue.poll();
            if (curr == null) {
                continue;
            }

            queue.add(curr.left);
            queue.add(curr.right);

            TreeNode left = curr.left;
            TreeNode right = curr.right;

            // Remove children from parent if children are unwanted
            if (curr.left != null && unwanted.contains(curr.left.val)) {
                curr.left = null;
            }
            if (curr.right != null && unwanted.contains(curr.right.val)) {
                curr.right = null;
            }

            // Add children to forest if parent is unwanted
            if (curr.left != null && unwanted.contains(curr.val)) {
                forest.add(left);
            }

            if (curr.right != null && unwanted.contains(curr.val)) {
                forest.add(right);
            }
        }

        List<TreeNode> ans = forest.stream().toList();
        return ans;
    }
}
// @leet end

File: 1110.delete-nodes-and-return-forest.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def delNodes(
        self, root: Optional[TreeNode], to_delete: List[int]
    ) -> List[TreeNode]:
        forest = []
        del_set = set(to_delete)

        def process_node(node: TreeNode) -> Optional[TreeNode]:
            if not node:
                return None

            node.left = process_node(node.left)
            node.right = process_node(node.right)

            if node.val in del_set:
                if node.left:
                    forest.append(node.left)
                if node.right:
                    forest.append(node.right)
                return None
            return node

        root = process_node(root)

        if root:
            forest.append(root)

        return forest


# @leet end
